### PR TITLE
Detach NamespaceStore from KubeWatchApi, proper store.loadAll()

### PR DESCRIPTION
**Pull request by ixrock**
- Detached `NamespaceStore` from `KubeWatchApi`
- Loading all namespaces by default with `KubeObjectStore.loadAll()`
- Added `KubeObjectStore.loadSelectedNamespaces()`

### CODE REVIEW

File: src/renderer/api/kube-watch-api.ts [hunk 1]
```diff
@@ -41,13 +40,13 @@ export interface IKubeWatchLog {
...
-  private cluster: Cluster;
-  private namespaceStore: NamespaceStore;
-
   private requestId = 0;
-  private isConnected = false;
   private reader: ReadableStreamReader<string>;
-  private subscribers = observable.map<KubeApi, number>();
+
+  @observable.ref private getCluster: () => Cluster;
+  @observable.ref private getNamespaces: () => string[];
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts]: These don't have an initial value so they should be marked as optional

```suggestion
  @observable.ref private getCluster?: () => Cluster;
  @observable.ref private getNamespaces?: () => string[];
```

File: src/renderer/api/kube-watch-api.ts [hunk 2]
```diff
@@ -57,16 +56,13 @@ export class KubeWatchApi {
...
-      if (!cluster.isAllowedResource(api.kind)) {
+    return Array.from(this.subscribers.keys()).map(api => {
+      if (!this.getCluster?.().isAllowedResource(api.kind)) {
         return [];
       }
 
       if (api.isNamespaced) {
-        return namespaceStore.getContextNamespaces().map(namespace => api.getWatchUrl(namespace));
+        return this.getNamespaces().map(namespace => api.getWatchUrl(namespace));
       } else {
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts]: If the getter isn't defined we should either throw a defined error or return `[]` instead of having the error be "cannot call undefined"

File: src/renderer/api/kube-watch-api.ts [hunk 3]
```diff
@@ -77,14 +73,16 @@ export class KubeWatchApi {
...
-    const { getHostedCluster } = await import("../../common/cluster-store");
-    const { namespaceStore } = await import("../components/+namespaces/namespace.store");
+  setupCluster(getter: () => Cluster) {
+    this.getCluster = getter;
+  }
 
-    await namespaceStore.whenReady;
+  setupWatchingNamespaces(getter: () => string[]) {
+    this.getNamespaces = getter;
+  }
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts]: imo these should be setters

File: src/renderer/api/kube-watch-api.ts [hunk 4]
```diff
@@ -143,7 +141,7 @@ export class KubeWatchApi {
     }
 
     if (preload) {
-      loading.push(...stores.map(store => store.loadAll()));
+      loading.push(...stores.map(store => store.loadAll(this.getNamespaces())));
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts]: `getNamespaces` might be undefined here

File: src/renderer/api/kube-watch-api.ts (hunk 2 above)
Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts]: Maybe, but it's used only internally and not critical imho.

File: src/renderer/api/kube-watch-api.ts (hunk 4 above)
Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts]: Yes, will fix.

**Comment by ixrock** [comment]
@Nokel81 please re-review again

File: src/renderer/api/kube-watch-api.ts (hunk 1 above)
Comment (Nokel81) [reply src/renderer/api/kube-watch-api.ts]: This was not changed or commented on, unresolving.

File: src/renderer/api/kube-watch-api.ts [hunk 5]
```diff
@@ -43,50 +42,59 @@ export interface IKubeWatchLog {
...
-  // events
   public onMessage = new EventEmitter<[IKubeWatchMessage]>();
 
+  @observable.ref private getCluster: () => Cluster;
+  @observable.ref private getNamespaces: () => string[];
+  @observable isConnected = false;
+  @observable subscribers = observable.map<KubeApi, number>();
+
+  @computed get isReady(): boolean {
+    return !!(this.cluster && this.namespaces);
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts]: I have generally come to prefer `Boolean(...)` vs `!!...` especially in places like this that are encapsulation points.

File: src/renderer/api/kube-watch-api.ts [hunk 6]
```diff
@@ -43,50 +42,59 @@ export interface IKubeWatchLog {
...
     return this.apis.length > 0;
   }
 
   @computed get apis(): string[] {
-    const { cluster, namespaceStore } = this;
-    const activeApis = Array.from(this.subscribers.keys());
-
-    return activeApis.map(api => {
-      if (!cluster.isAllowedResource(api.kind)) {
+    return Array.from(this.subscribers.keys()).map(api => {
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts:67]: I recently learned about this, which I think is probably faster since it doesn't have to create the intermediate array.
```suggestion
    return Array.from(this.subscribers.keys(), api => {
```

File: src/renderer/api/kube-watch-api.ts [hunk 7]
```diff
@@ -43,50 +42,59 @@ export interface IKubeWatchLog {
...
+      if (!this.isReady || !this.isAllowedApi(api)) {
```
Comment (Nokel81) [inline src/renderer/api/kube-watch-api.ts]: the `!this.isReady` check should probably be hoisted of the map function.

File: src/renderer/api/kube-watch-api.ts (hunk 6 above)
Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts:67]: I believe that such micro-optimizations we should avoid in advance. In most cases you won't notice any performance difference..
If it brings more readability or reasonable performance boost, only then apply (but _it is _not__ in that case + here you gets increased cognitive load for rare-used second argument of `Array.from`).

File: src/renderer/api/kube-watch-api.ts (hunk 7 above)
Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts:72]: Why?

Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts]: done

File: src/renderer/api/kube-watch-api.ts (hunk 5 above)
Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts]: done

File: src/renderer/api/kube-watch-api.ts (hunk 1 above)
Comment (ixrock) [reply src/renderer/api/kube-watch-api.ts]: fixed, PTAL

File: src/main/routes/watch-route.ts [hunk 8]
```diff
@@ -65,50 +90,72 @@ class ApiWatcher {
...
 
       return;
     }
 
-    response.setHeader("Content-Type", "text/event-stream");
+    this.setResponse(response);
+    response.setHeader("Content-Type", "application/json");
     response.setHeader("Cache-Control", "no-cache");
     response.setHeader("Connection", "keep-alive");
     logger.debug(`watch using kubeconfig:${JSON.stringify(cluster.getProxyKubeconfig(), null, 2)}`);
```
Comment (Nokel81) [inline src/main/routes/watch-route.ts]: Logger already has this functionality and logging should be single lines
```suggestion
    logger.debug("watch using kubeconfig", { kubeconfig: cluster.getProxyKubeconfig() });
```

**Comment by Nokel81** [comment]
@ixrock Please rebase (and maybe even squash) this against master. It has several conflicting versions (compared to master at least) which makes it hard to review. Thanks.

File: src/renderer/components/app.tsx [hunk 9]
```diff
@@ -177,7 +177,7 @@ export class App extends React.Component {
       }
     });
 
-    for (const menu of this.extensionRoutes.keys()) {
```
Comment (Nokel81) [inline src/renderer/components/app.tsx]: Why was this changed? It is both slower and more verbose.

Comment (ixrock) [reply src/renderer/components/app.tsx]: Wrong merge, you're faster than my revert-fix :D

Comment (Nokel81) [reply src/renderer/components/app.tsx]: lol

**Comment by ixrock** [comment]
rebased to https://github.com/lensapp/lens/pull/2063
